File: 1/day1.py
with open('input.txt', 'r', encoding='utf-8') as f:
    puzzle: str = f.read()

locids: list[tuple[int, int]] = [
    tuple(map(int, i.split())) for i in puzzle.split('\n') \
    if i.strip() != ''
]

left: list[int] = sorted(i[0] for i in locids)
right: list[int] = sorted(i[1] for i in locids)

dist: int = 0
simscore: int = 0

for a, b in zip(left, right):
    dist += abs(a - b)
    simscore += a * right.count(a)

print('Day 1, Part 1')
print(dist)
print('Day 1, Part 2')
print(simscore)

File: 1/haskell/day1.hs
import Data.List

count :: Eq a => [a] -> a -> Int
count lst target = length $ filter (== target) lst

main = do
    puzzle <- readFile "../input.txt"
    let locids::[[Int]] = map (map read . words) . lines $ puzzle
    let left::[Int] = sort $ map head locids
    let right::[Int] = sort $ map (!! 1) locids
    let dist::Int = sum $ map abs $ zipWith (-) left right
    let simscore::Int = sum $ map (\i -> i * count right i) left

    print "Day 1, Part 1"
    print dist
    print "Day 1, Part 2"
    print simscore
